### DISCUSSION TITLE
Wakatime support

### DISCUSSION THREAD

**Discussion by joetifa2003** [helix-editor/helix#3477] [Ideas]
Is it planned to have Wakatime support, and is there any workaround to get it working?

**Comment by archseer** [comment]
It's not planned and should be an external plugin. You could implement it in a fork for the time being

**Comment by AceofSpades5757** [comment]
For reference, here are some links.

[WakaTime Dev Docs](https://wakatime.com/developers)
[WakaTime GitHub](https://github.com/wakatime)

**Comment by adenoz** [comment]
Just bumping this. Just starting to use Helix properly now and would really miss not having wakatime. Is anyone working on a potential plug-in?

Edit: I now realise the plug-in system is still some way away. Regardless, I've raised an issue in the unofficial Helix plugin ideas repo [here](https://github.com/zim0369/helix_plugins/issues/6).

**Comment by n3oney** [comment]
I've created a workaround that works for me!

A WakaTime plugin pretty much just needs to call the wakatime-cli every once in a while with the filename passed. I figured, since I already save files quite often, and I have auto-formatting enabled, I'll write a wrapper script around the formatter that apart from the formatter, also calls the cli!

(Note - this needs https://github.com/helix-editor/helix/pull/5626)

The script looks something like this:

```
#!/bin/bash
prettierd "$1" 2>/dev/null
wakatime-cli --entity "$1" --key "YOUR_WAKATIME_KEY_OR_DONT_SPECIFY_SO_IT_READS_FROM_WAKATIME_CONFIG_FILE" --plugin "Helix/version Helix/version" --write 2>&1 >/dev/null &
```

I use it like this:

![image](https://github.com/helix-editor/helix/assets/30625554/7975da79-35cb-440a-ae11-b14180d88711)

It has been working so far.

**Reply by noahfraiture** [reply]
Hey, it doesn't work for me
The problem is that my config actually don't use the bash file
```
[[language]]
name = "rust"
language-servers = ["rust-analyzer" ]
auto-format = true
[language.formatter]
args = ["{}"]
command = "/home/noah/.config/helix/wakatime.sh"
[language-server.rust-analyzer]
command = "rust-analyzer"
[language-server.rust-analyzer.config.check]
command = "clippy"
```
wakatime.sh
```
#!/bin/bash
echo "empty$1" >/home/noah/t && wakatime-cli --entity "$1" --plugin "Helix/version Helix/version" --write 2>&1 >/dev/null &
rustfmt
```

I put an echo "thing">file in the bash file to see if the file is well execute and it isn't. Did I do something wrong ? My file is well formatted but it's well formatted even if I remoe the file actually

**Reply by n3oney** [reply]
I have no clue, as I actually stopped using this method after some time, because the times were inaccurate.

**Reply by noahfraiture** [reply]
I finally made my self but it requires to merge an unmerged PR (https://github.com/helix-editor/helix/pull/5626) to be able to specify the file in a command.
I have this bash file 
```#!/bin/bash
echo "[$(date +%s)] wakatime send : $1" >>/home/noah/.config/helix/wakatime.log && 
wakatime-cli --entity "$1" --plugin "Helix/version Helix/version" --write 2>&1 >/dev/null &&

for i in {1..3}
do
echo "[$(date +%s)] wakatime send : $1. Iteration $i" >>/home/noah/.config/helix/wakatime.log && 
  wakatime-cli --entity "$1" --plugin "Helix/$(hx --version | cut -d' ' -f2) Helix-wakatime/$(hx --version | cut -d' ' -f2)" 2>&1 >/dev/null
  sleep 120
done
```
And this configuration : 
```
[keys.normal]
C-s = [":sh bash ~/.config/helix/wakatime.sh %{filename}", ":w"]
```

However I still have the problem of the editor not recognized

**Reply by n3oney** [reply]
Yep, as I mentioned originally requires that PR

**Comment by mrnossiom** [comment]
I made https://github.com/mrnossiom/wakatime-lsp. The bulk of the work (sending code heartbeats) is done by WakaTime's CLI, but it's notified via LSP events (e.g. change, save, etc.) Might want to check it out. I tried to make a nice README.

**Reply by noahfraiture** [reply]
This is probably the best options since the PR https://github.com/helix-editor/helix/pull/6979 is not merge for reason I don't know

**Reply by kirawi** [reply]
That PR has some issues with parsing IIRC, @pascalkuthe
